File: dotnet/EliteDangerousSdk/Planner/EngineeringPlanner.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using EliteDangerousSdk.Data;

namespace EliteDangerousSdk.Planner
{
    public class PlannedModification
    {
        public string ModuleGroup { get; set; } = "";
        public string BlueprintName { get; set; } = "";
        public int Grade { get; set; } = 1;
        public string? ExperimentalEffect { get; set; }
    }

    public class MaterialCost
    {
        public string Material { get; set; } = "";
        public int Quantity { get; set; }
        public int Grade { get; set; }
        public string Source { get; set; } = "blueprint";
    }

    public class EngineerVisit
    {
        public string Engineer { get; set; } = "";
        public string ModuleGroup { get; set; } = "";
        public string BlueprintName { get; set; } = "";
        public int Grade { get; set; }
    }

    public class EngineeringPlan
    {
        public List<MaterialCost> Materials { get; set; } = new();
        public Dictionary<string, int> MaterialTotal { get; set; } = new();
        public List<string> Engineers { get; set; } = new();
        public List<EngineerVisit> EngineerVisits { get; set; } = new();
    }

    public static class EngineeringPlanner
    {
        public static EngineeringPlan PlanEngineering(List<PlannedModification> modifications)
        {
            var materials = new List<MaterialCost>();
            var engineerVisits = new List<EngineerVisit>();
            var engineerSet = new HashSet<string>();

            var blueprints = DataProvider.Blueprints;
            var specials = DataProvider.Specials;
            var moduleBpMap = DataProvider.ModuleBlueprintMap;

            foreach (var mod in modifications)
            {
                // Blueprint material costs
                if (blueprints.ValueKind == JsonValueKind.Object &&
                    blueprints.TryGetProperty(mod.BlueprintName, out var bp))
                {
                    var grades = bp.GetProperty("grades");
                    if (grades.TryGetProperty(mod.Grade.ToString(), out var gradeData) &&
                        gradeData.TryGetProperty("components", out var components))
                    {
                        foreach (var comp in components.EnumerateObject())
                        {
                            materials.Add(new MaterialCost
                            {
                                Material = comp.Name,
                                Quantity = comp.Value.GetInt32(),
                                Grade = mod.Grade,
                                Source = "blueprint",
                            });
                        }
                    }
                }

                // Experimental effect materials
                if (!string.IsNullOrEmpty(mod.ExperimentalEffect) &&
                    specials.ValueKind == JsonValueKind.Object &&
                    specials.TryGetProperty(mod.ExperimentalEffect, out var special) &&
                    special.TryGetProperty("components", out var expComponents))
                {
                    foreach (var comp in expComponents.EnumerateObject())
                    {
                        materials.Add(new MaterialCost
                        {
                            Material = comp.Name,
                            Quantity = comp.Value.GetInt32(),
                            Grade = 0,
                            Source = "experimental",
                        });
                    }
                }

                // Engineer lookup
                if (moduleBpMap.ValueKind == JsonValueKind.Object &&
                    moduleBpMap.TryGetProperty(mod.ModuleGroup, out var mapping) &&
                    mapping.TryGetProperty("blueprints", out var bps) &&
                    bps.TryGetProperty(mod.BlueprintName, out var bpEntry) &&
                    bpEntry.TryGetProperty("grades", out var gradeMap) &&
                    gradeMap.TryGetProperty(mod.Grade.ToString(), out var engGrade) &&
                    engGrade.TryGetProperty("engineers", out var engineersArr))
                {
                    foreach (var eng in engineersArr.EnumerateArray())
                    {
                        var engName = eng.GetString() ?? "";
                        if (engineerSet.Add(engName))
                        {
                            engineerVisits.Add(new EngineerVisit
                            {
                                Engineer = engName,
                                ModuleGroup = mod.ModuleGroup,
                                BlueprintName = mod.BlueprintName,
                                Grade = mod.Grade,
                            });
                        }
                    }
                }
            }

            var materialTotal = new Dictionary<string, int>();
            foreach (var m in materials)
            {
                materialTotal.TryGetValue(m.Material, out var existing);
                materialTotal[m.Material] = existing + m.Quantity;
            }

            return new EngineeringPlan
            {
                Materials = materials,
                MaterialTotal = materialTotal,
                Engineers = engineerSet.OrderBy(e => e).ToList(),
                EngineerVisits = engineerVisits,
            };
        }

        public static List<MaterialCost> GetBlueprintComponents(string blueprintName, int grade)
        {
            var result = new List<MaterialCost>();
            var blueprints = DataProvider.Blueprints;

            if (blueprints.ValueKind != JsonValueKind.Object ||
                !blueprints.TryGetProperty(blueprintName, out var bp)) return result;

            var grades = bp.GetProperty("grades");
            if (!grades.TryGetProperty(grade.ToString(), out var gradeData)) return result;
            if (!gradeData.TryGetProperty("components", out var components)) return result;

            foreach (var comp in components.EnumerateObject())
            {
                result.Add(new MaterialCost
                {
                    Material = comp.Name,
                    Quantity = comp.Value.GetInt32(),
                    Grade = grade,
                    Source = "blueprint",
                });
            }
            return result;
        }

        public static List<MaterialCost> GetExperimentalEffectComponents(string effectName)
        {
            var result = new List<MaterialCost>();
            var specials = DataProvider.Specials;

            if (specials.ValueKind != JsonValueKind.Object ||
                !specials.TryGetProperty(effectName, out var special)) return result;
            if (!special.TryGetProperty("components", out var components)) return result;

            foreach (var comp in components.EnumerateObject())
            {
                result.Add(new MaterialCost
                {
                    Material = comp.Name,
                    Quantity = comp.Value.GetInt32(),
                    Grade = 0,
                    Source = "experimental",
                });
            }
            return result;
        }

        public static List<string> GetEngineersForBlueprint(string moduleGroup, string blueprintName, int grade)
        {
            var result = new List<string>();
            var moduleBpMap = DataProvider.ModuleBlueprintMap;

            if (moduleBpMap.ValueKind != JsonValueKind.Object) return result;
            if (!moduleBpMap.TryGetProperty(moduleGroup, out var mapping)) return result;
            if (!mapping.TryGetProperty("blueprints", out var bps)) return result;
            if (!bps.TryGetProperty(blueprintName, out var bpEntry)) return result;
            if (!bpEntry.TryGetProperty("grades", out var gradeMap)) return result;
            if (!gradeMap.TryGetProperty(grade.ToString(), out var engGrade)) return result;
            if (!engGrade.TryGetProperty("engineers", out var engineersArr)) return result;

            foreach (var eng in engineersArr.EnumerateArray())
            {
                var engName = eng.GetString();
                if (engName != null) result.Add(engName);
            }
            return result;
        }
    }
}

File: core/packages/planner/src/engineering-planner.ts
import {
  blueprints as blueprintsData,
  moduleBlueprintMappings as bpMappings,
  specials as specialsData,
} from "@elite-dangerous-sdk/data";

export interface PlannedModification {
  moduleGroup: string;
  blueprintName: string;
  grade: number;
  experimentalEffect?: string;
}

export interface MaterialCost {
  material: string;
  quantity: number;
  grade: number;
  source: "blueprint" | "experimental";
}

export interface EngineerVisit {
  engineer: string;
  moduleGroup: string;
  blueprintName: string;
  grade: number;
}

export interface EngineeringPlan {
  materials: MaterialCost[];
  materialTotal: Record<string, number>;
  engineers: string[];
  engineerVisits: EngineerVisit[];
}

export function planEngineering(
  modifications: PlannedModification[],
): EngineeringPlan {
  const materials: MaterialCost[] = [];
  const engineerVisits: EngineerVisit[] = [];
  const engineerSet = new Set<string>();

  for (const mod of modifications) {
    const bp = (blueprintsData as Record<string, any>)[mod.blueprintName];
    const gradeKey = String(mod.grade);
    const gradeData = bp?.grades?.[gradeKey];

    if (gradeData?.components) {
      for (const [mat, qty] of Object.entries(
        gradeData.components as Record<string, number>,
      )) {
        materials.push({
          material: mat,
          quantity: qty,
          grade: mod.grade,
          source: "blueprint",
        });
      }
    }

    if (mod.experimentalEffect) {
      const special = (specialsData as Record<string, any>)[
        mod.experimentalEffect
      ];
      if (special?.components) {
        for (const [mat, qty] of Object.entries(
          special.components as Record<string, number>,
        )) {
          materials.push({
            material: mat,
            quantity: qty,
            grade: 0,
            source: "experimental",
          });
        }
      }
    }

    const mapping = (bpMappings as Record<string, any>)[mod.moduleGroup];
    const bpEntry = mapping?.blueprints?.[mod.blueprintName];
    const engineers = bpEntry?.grades?.[gradeKey]?.engineers as
      | string[]
      | undefined;

    if (engineers) {
      for (const eng of engineers) {
        engineerSet.add(eng);
        engineerVisits.push({
          engineer: eng,
          moduleGroup: mod.moduleGroup,
          blueprintName: mod.blueprintName,
          grade: mod.grade,
        });
      }
    }
  }

  const materialTotal: Record<string, number> = {};
  for (const m of materials) {
    materialTotal[m.material] = (materialTotal[m.material] ?? 0) + m.quantity;
  }

  return {
    materials,
    materialTotal,
    engineers: [...engineerSet].sort(),
    engineerVisits,
  };
}

export function getBlueprintComponents(
  blueprintName: string,
  grade: number,
): MaterialCost[] {
  const bp = (blueprintsData as Record<string, any>)[blueprintName];
  const gradeKey = String(grade);
  const gradeData = bp?.grades?.[gradeKey];
  if (!gradeData?.components) return [];

  return Object.entries(gradeData.components as Record<string, number>).map(
    ([material, quantity]) => ({
      material,
      quantity,
      grade,
      source: "blueprint" as const,
    }),
  );
}

export function getExperimentalEffectComponents(
  effectName: string,
): MaterialCost[] {
  const special = (specialsData as Record<string, any>)[effectName];
  if (!special?.components) return [];

  return Object.entries(special.components as Record<string, number>).map(
    ([material, quantity]) => ({
      material,
      quantity,
      grade: 0,
      source: "experimental" as const,
    }),
  );
}

export function getEngineersForBlueprint(
  moduleGroup: string,
  blueprintName: string,
  grade: number,
): string[] {
  const mapping = (bpMappings as Record<string, any>)[moduleGroup];
  const bpEntry = mapping?.blueprints?.[blueprintName];
  const engineers = bpEntry?.grades?.[String(grade)]?.engineers;
  return engineers ? [...engineers] : [];
}
